src/librustdoc/passes/collect_intra_doc_links.rs

```diff
@@ -1089,10 +1107,61 @@ impl LinkCollector<'_, '_> {
         ori_link: &str,
         link_range: Option<Range<usize>>,
     ) -> Option<(Res, Option<String>)> {
+        let lookup_key = (disambiguator.map(Disambiguator::ns), ori_link.to_owned());
+
+        // Try to look up both the result and the corresponding side channel value
+        if let Some((res, side_channel)) = self.visited_links.get(&lookup_key) {
+            self.kind_side_channel.set(side_channel.clone());
+            return res.clone();
+        }
+
+        if let Some((res, ns)) = self.do_resolve_with_disambiguator(
+            disambiguator,
+            item,
+            dox,
+            path_str,
+            current_item,
+            base_node,
+            extra_fragment.clone(),
+            ori_link,
+            link_range,
+        ) {
+            if disambiguator.is_none() {
+                // Store result for the "look for all namespaces" case
+                self.visited_links.insert(
+                    lookup_key,
+                    (Some(res.clone()), self.kind_side_channel.clone().into_inner()),
+                );
+            }
+
+            // Store result for the actual namespace
+            let key = (Some(ns), ori_link.to_owned());
+            self.visited_links
+                .insert(key, (Some(res.clone()), self.kind_side_channel.clone().into_inner()));
+            Some(res)
+        } else {
+            // Also store a failed resolution
+            self.visited_links.insert(lookup_key, (None, None));
+            None
+        }
+    }
+
+    fn do_resolve_with_disambiguator(
```

Comment:
Rather than renaming this, I'd call your new function `resolve_with_disambiguator_cached` or something like that.

Reply:
Fair enough, thanks.